bpf/gotracer/go_sql.c

```diff
@@ -25,11 +25,119 @@
 #include <common/ringbuf.h>
 
 #include <gotracer/go_common.h>
+#include <gotracer/go_str.h>
 
-static __always_inline void set_sql_info(void *goroutine_addr, void *sql_param, void *query_len) {
+// Validates that driverConn.ci points to a MySQL connection and returns the mysqlConn pointer.
+static __always_inline void *get_mysql_conn_ptr(u64 driver_conn_ptr) {
+    if (driver_conn_ptr == 0) {
+        return NULL;
+    }
+
+    off_table_t *ot = get_offsets_table();
+
+    // Get driverConn.ci offset
+    u64 ci_offset = go_offset_of(ot, (go_offset){.v = _driverconn_ci_pos});
+    if (!ci_offset) {
+        bpf_dbg_printk("can't get driverConn.ci offset");
+        return NULL;
+    }
+
+    // driverConn.ci is a Go interface [type_ptr (8 bytes), data_ptr (8 bytes)]
+    // Read the type pointer (at ci_offset + 0) to validate driver type
+    void *ci_type_ptr = 0;
+    int res =
+        bpf_probe_read(&ci_type_ptr, sizeof(ci_type_ptr), (void *)(driver_conn_ptr + ci_offset));
```

Comment:
```suggestion
        bpf_probe_read_user(&ci_type_ptr, sizeof(ci_type_ptr), (void *)(driver_conn_ptr + ci_offset));
```

From the bpf helpers manpage:

```

       long bpf_probe_read(void *dst, u32 size, const void *unsafe_ptr)

              Description
                     For tracing programs, safely attempt to read size
                     bytes from kernel space address unsafe_ptr and store
                     the data in dst.

                     Generally, use bpf_probe_read_user() or
                     bpf_probe_read_kernel() instead.

              Return 0 on success, or a negative error in case of
                     failure.
```

The reason is because `bpf_probe_read` needs an extra step to resolve if the ptr in question points to user or kernel memory. It's useful  in common code that handles both kernel and userspace pointers, but here I believe it's all user memory (correct me if I am wrong) - so `bpf_probe_read_user` is more appropriate (and also tells the reader the nature of the pointer).